internal/pkg/addon/package.go

```diff
@@ -0,0 +1,243 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package addon
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/aws/copilot-cli/internal/pkg/aws/s3"
+	"gopkg.in/yaml.v3"
+)
+
+type uploader interface {
+	Upload(bucket, key string, data io.Reader) (string, error)
+}
+
+// packagePropertyConfig defines how to package a particular property in a cloudformation resource.
+// There are two ways replacements occur. Given a resource configuration like:
+//  MyResource:
+//    Type: AWS::Resource::Type
+//    Properties:
+//      <Property>: file/path
+//
+// Without BucketNameProprty and ObjectKeyProperty, `file/path` is directly replaced with
+// the S3 location the contents were uploaded to, resulting in this:
+//  MyResource:
+//    Type: AWS::Resource::Type
+//    Properties:
+//      <Property>: s3://bucket/hash
+//
+// If BucketNameProperty and ObjectKeyProperty are set, the value of <Property> is changed to a map
+// with BucketNameProperty and ObjectKeyProperty as the keys.
+//  MyResource:
+//    Type: AWS::Resource::Type
+//    Properties:
+//      <Property>:
+//        <BucketNameProperty>: bucket
+//        <ObjectKeyProperty>: hash
+type packagePropertyConfig struct {
+	// Property is the key in a cloudformation resource's 'Properties' map to be packaged.
+	// Nested properties are represented by multiple keys in the slice, so the field
+	//  Properties:
+	//    Code:
+	//      S3: ./file-name
+	// is represented by []string{"Code", "S3"}.
+	Property []string
+
+	// BucketNameProperty represents the key in a submap of Property, created
+	// after uploading an asset to S3. If this and ObjectKeyProperty are empty,
+	// a submap will not be created and an S3 location URI will replace value of Property.
+	BucketNameProperty string
+
+	// ObjectKeyProperty represents the key in a submap of Property, created
+	// after uploading an asset to S3. If this and BucketNameProperty are empty,
+	// a submap will not be created and an S3 location URI will replace value of Property.
+	ObjectKeyProperty string
+
+	// ForceZip will force a zip file to be created even if the given file URI
+	// points to a file. Directories are always zipped.
+	ForceZip bool
+}
+
+// resourcePackageConfig maps a CloudFormation resource type to configuration
+// for how to transform it's properties.
+//
+// TODO(dnrnd) AWS::Include.Location
```

Comment:
yeah - just adding it in a future PR! since it can appear in more places - not just properties of resources - it's going to be processed differently.

---

I was planning on adding `AWS:Serverless` resources (they'd be configured the same way), but in my testing with `AWS::Serverless::Function` in an addon, I always get the error:
```
 Template format error: Unrecognized resource types: [AWS::Serverless::Function]
```
Similarly, I also get the same error with a `AWS::Serverless::Function` while using the aws cli to run `cloudformation deploy`. From that, I _think_ the `AWS::Serverless` resources only work when using SAM CLI. (https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-specification.html)

Reply:
oh! It should work if you add the following to your template:
```
Transform: AWS::Serverless-2016-10-31
```
CFN will then run a macro behind the scenes to transform the template and expand the resources into regular service resources

Reply:
oh🙄 then yes i will add them :joy: 

(also just tested and you're right, it works 😊)